src/renderer/vt/VtSequences.cpp

```diff
@@ -248,11 +249,10 @@ using namespace Microsoft::Console::Render;
 [[nodiscard]] HRESULT VtEngine::_SetGraphicsRendition256Color(const WORD index,
                                                               const bool fIsForeground) noexcept
 {
-    const std::string fmt = fIsForeground ?
-                                "\x1b[38;5;%dm" :
-                                "\x1b[48;5;%dm";
-
-    return _WriteFormattedString(&fmt, ::Xterm256ToWindowsIndex(index));
+    // An example string with max length would be "\x1b[38;5;128m", which has length = 11.
+    char buf[11];
+    char* end = fmt::format_to(std::begin(buf), FMT_COMPILE("\x1b[{};5;{}m"), fIsForeground ? 38 : 48, ::Xterm256ToWindowsIndex(index));
+    return _Write({ buf, static_cast<std::string_view::size_type>(end - buf) });

```

Comment:
I'm strongly concerned with the memory safety of this code.
First of all you forgot to include the 0 byte in your buffer size calculation. I'm not trying to diss you here, rather this simply shows how such things are best abstracted away behind safer container classes. Even if the `buf` was 12 bytes large, it would fail the moment anyone touches the code and the reviewer forgets to re-check the correctness of the buffer size.
Additionally `Xterm256ToWindowsIndex` can theoretically return values as large as 2^16-1, resulting in a buffer size of 14!

MSVC's small string optimization (SSO) is unfortunately among the worst of all C++ stdlib implementations. But regardless its 15 byte buffer is plenty enough for both `_SetGraphicsRendition256Color` and `_SetGraphicsRendition16Color`, and allows you to use the regular `fmt::format` function without incurring any allocations.

👉 I'm convinced you should use `fmt::format()` here.

Reply:
Thanks Leonard for the detailed explanation. I did try to have the \0 at first , then I changed my implementation from char * to std::string, and then changed it back, to test the performance under various conditions. Finally l got lost, so is the \0.

I totally understand the concern about memory safety, which is my biggest concern. But as Dustin pointed out  in https://github.com/microsoft/terminal/pull/10426#issuecomment-861851772, fmt::format simply is not performant enough. Will post a trace later.

Reply:
Actually I‘m totally wrong about the 0 byte!
Since fmt expects an iterator I’m sure it doesn’t manually append a 0 byte at the end but rather let’s this be done by the container (for instance std::string).
So your Code was actually just fine! 😅